action/protocol/poll/protocol.go

```diff
@@ -62,8 +62,8 @@ type ProductivityByEpoch func(context.Context, uint64) (uint64, map[string]uint6
 type Protocol interface {
 	protocol.Protocol
 	protocol.GenesisStateCreator
-	DelegatesByEpoch(context.Context, uint64) (state.CandidateList, error)
-	CandidatesByHeight(context.Context, uint64) (state.CandidateList, error)
+	DelegatesByEpoch(context.Context, protocol.StateReader, uint64) (state.CandidateList, error)
```

Comment:
it is possible to kill the last parameter "epochNum"?

Reply:
yes it is possible, but we have to compare epochnumber with tipEpochnumber not inside of  DelegatesByEpoch but outside of it which happens in consensus/productivityByEpoch/ blockchain.context(). Is it ok?

Reply:
introduce a function "Delegates" to distinguish the historical read and tip read. If necessary, you can have one more function NextDelegates

Reply:
ok! I changed to have Delegates() and NextDelegates() function. Btw, we don't have to distinguish historical and tip read, because now DelegatesByEpoch is only for tip